submissions/issue-20/milestone-1/fee1-dead/20210919-milestone-1.md

```diff
@@ -0,0 +1,119 @@
+be reconsidered when it turns out otherwise.
+
+Then, the crate defines a wrapper type `FlowClient` which provides helper functions
+for `GrpcClient` implementations. A helper functions can be written as:
+
+```rust
+impl<Inner> FlowClient<Inner> {
+    pub fn fn_name<'a>(&'a mut self, args..) -> Inner::Output
+        where &'a mut Inner: GrpcClient<InputTy, OutputTy> 
+    {
+        self.send(/* Request Construct */)
+    }
+} 
+```
+
+With the help of a macro, it becomes very simple to add new functions: 
+([src](https://github.com/fee1-dead/flow.rs/blob/82510b45d3b14b179192a5d2bccec87c932d7819/src/client.rs#L67-L77))
+
+```rust
+define_reqs! {
+    /// Shortcut for `self.send(PingRequest {})`.
+    pub fn ping() PingRequest => PingResponse {
+        PingRequest {}
+    }
+
+    /// Returns information of the latest block.
+    pub fn latest_block_header(is_sealed: bool)
+        GetLatestBlockHeaderRequest => BlockHeaderResponse
+    {
+        GetLatestBlockHeaderRequest { is_sealed }
+    }
+}
+```
+
+A `FlowClient` instance can be easily created, via the `tonic` and `hyper` crate:
+
+```rust
+pub type TonicFlowClient<Service> = FlowClient<Grpc<Service>>;
+pub type TonicHyperFlowClient = TonicFlowClient<Channel>;
+
+impl TonicHyperFlowClient {
+    pub fn mainnet() -> Result<Self, tonic::transport::Error> {
+        Ok(Self {
+            inner: Grpc::new(
+                Channel::from_static("http://access.mainnet.nodes.onflow.org:9000").connect_lazy()?,
+            ),
+        })
+    }
+}
+```
+
+## User Stories
+
+Most user stories can be implemented with associated functions for `FlowClient`s.
+i.e. a function that the user can call through the Access API.
+
+Some will require parsing JSON-Cadence Data. (such as "submit a script and parse
+the response") I plan to do this with `serde-json` or a crate providing similar
+functionality (probably split to a new crate that parses JSON-Cadence)
+
+Some will require signing, which will use the `k256` crate or another better crate
+with the same functionality.
+
+Blocks:
+- retrieve a block by ID - Access API
+- retrieve a block by height - Access API
+- retrieve the latest block - Access API
+
+Collections:
+- retrieve a collection by ID - Access API
+
+Events:
+- retrieve events by name in the block height range - Access API
+
+Scripts:
+- submit a script and parse the response - Needs parsing
+- submit a script with arguments and parse the response - Needs parsing
+
+Accounts:
+- retrieve an account by address - Access API
+- create a new account - A transaction with provided cadence template
+- deploy a new contract to the account - A transaction with provided cadence template
+- remove a contract from the account - A transaction with provided cadence template
+- update an existing contract on the account - A transaction with provided cadence template
+
+Transactions: 
+- retrieve a transaction by ID - Access API
```

Comment:
Retrieving a transaction will also require JSON-Cadence parsing, since a transaction contains arguments that are encoded as JSON-Cadence.